### ISSUE TITLE
Make the recompute -> execute operation use multiple processors

### ISSUE THREAD

**Issue by bgbsww** [FreeCAD/FPA-grant-proposals#14]
### Proposal description

FreeCAD suffers from being essentially a single threaded program that is unable to take advantage of multiple cores on a modern CPU.  I propose to remove part of this restriction and provide a framework for future multithreading innovations by creating a multithreading system within FreeCAD, and then extending the recompute -> execute operation to allow multithreading.  This single adjustment should provide a significant performance and responsiveness improvement to FreeCAD.

This work will be carried out in phases, both for visibility and to allow for early exits or redirects if warranted.  The initial phases are research, the middle phases are implementation, and the final phases are documentation.

### Deliverables

1. Concise report on the potential concurrency options for FreeCAD; multiprocessing, and multithreading using std, Boost, Qt or other.
2. Concise report on implications for the Boost and Qt signal/slot systems.
3. Concise report on thread safety for components (Boost, Qt, OCCT, others).
4. Research replacements for Document::getDependencyList that return independent sets of DocumentObjects suitable for parallel processing.
5. Technical proposal for FC process or thread layout, and implementation technology.
6. Gathering consensus on proposal.
7. Preliminary profiling to generate baseline and estimate speedup.
8. Implement changes to create parallelizable segmented topoSortedObjects list for recompute.
9. Implement changes to support multiprocess / multithreading.
10. Implement changes for multiprocess / multithreading support of recompute / execute.
11. Final Profiling.
12. Technical documentation for use of multiprocessing / multithreading in other projects.

### Timeline

1.  Week 1
2.  Week 2
3.  Week 3
4.  Week 4
5.  Week 5
6.  Weeks 6 & 7
7.  Week 6
8.  Week 8
9.  Week 9
10.  Week 10
11.  Week 11
12.  Week 11

### Risks and mitigation

The implementation portion can be cancelled if the research and consensus phases fail.  The research portion is likely a useful outcome on its own if it unveils deeper problems.

The sorting / splitting mechanism for the DocumentObjects must be performant and effective.

The mechanisms chosen must not effect the stability of the application.

The mechanisms chosen must not significantly increase the system resource required to run the application in a single process/thread mode.

The mechanisms chosen must not significantly increase the complexity of developing in and against FreeCAD.

### Compensation

$5000 for phases 1-4
$2000 for phase 5
$0 for phase 6
$5000 for phases 7-10
$5000 for phases 11-12

### About you

Bradley McLean /  bgbsww @gmail  on github, on telegram, on discord.

I am a frequent FreeCAD contributor, hopefully a known quantity in the community.  I have worked on Toponaming, OCCT version porting and generaly debugging across FreeCAD.  I have past experience in application, system, web and microcontroller coding.  I've worked on various multiprocess and multithreading software.

**Comment by WandererFan** [comment]
FYI.  There are a number places in the codebase that use the QtConcurrent mechanism.

OpenCascade is only thread safe in places.  Generally you can't reliably update a single OpenCascade object from multiple threads.  

Would love to see the Gui in its own thread to avoid lock out during long operations.

**Comment by bgbsww** [comment]
Thank-you!  This is my current understanding, subject to correction (as are many things in this project):

1)  Using QtConcurrent as a general mechanism would "leak" the GUI into the App in various places, and so would need to be carefully encapsulated as we did it, and we would have to accept the Qt dependency in non GUI executables.  I think this is to be avoided, and so the solution for App must co-exist with the use of QtConcurrent in the Gui code only.  Or, I suppose replacing the QtConcurrent use, but I'm not a fan of that.  I am initially biased towards a std::thread solution for App/main, but need to prove it is the best choice.

2) I acknowledge that thread safety of OpenCascade will be a moving target that should not be depended upon.  It appears that many (most?  all?) of our uses of OpenCascade involve a readily encapsulatable set of calls that require no prior or subsequent state.  So the challenge is to keep the set of calls together on a single thread.  That seems reasonable.

3) A naive first pass would place App in a thread [the main default thread], Gui in a set of QtConcurrent threads, I.E. adding one thread to the current setup, and OpenCascade in a pool of threads.  Whether any of these make sense will require investigation in each case as to performance impact, code impact, and likely other things I haven't anticipated.  Gui in its own thread [pool?] is necessary for but not sufficient to a good user experience, and I do not intend to tackle all the related changes like activity spinners, operation aborts, semi-modal behavior, etc.  This is just the underlying mechanisms and the one (big) example case of the recompute. 

You'll note that this leans multithreaded as opposed to multiprocess.  I (pre-research) think that's the right approach here, but I am keeping an open mind if it turns out to be impractical.

**Comment by WandererFan** [comment]
QtConcurrent is already used on the App side, at least in TD.  I'd have to look again, but I think it is used on the App side in other modules also.

**Comment by bgbsww** [comment]
It's in Mesh, among other places.  Okay, that might tweak the approach a bit, but we're getting into the research phase of the project.  Thanks!

**Comment by yorikvanhavre** [comment]
This is of course a gold target and without a doubt something interesting for the FPA to fund.

I have a concern over your proposed timeline, only. It seems to me such an adventure passes through large chunks of unknown, an the solutions that emerge cannot be put into action without general support of the developers community. And this might take time, possibly much more than 11 weeks.

Would your plan still be valid for you if it takes 6 months? 1 year maybe? Just imagining worst case scenarios here :)

**Comment by bgbsww** [comment]
Good point.  Yes, I think the plan is still valid if it takes longer.  There is obviously a risk that any given stage runs long, and a risk that the time commitment becomes a problem.  The phasing of the project and payments are intended to reduce the exposure for the FPA.

I'll also acknowledge the possibility that the plan might be forced to evolve, in which case further discussion of later stages  might have to take place.

I have, however done enough early research and test implementation to feel comfortable with what is laid out here as reasonable given what I know now.

**Comment by chennes** [comment]
The FPA grant review committee has reviewed this application, and recommends that it be funded. FPA voting will commence today, and you can expect it to take about two weeks. Some comments from reviewers:

“This task is quite challenging but has the potential to significantly advance FreeCAD. While it’s not something I would assign to just any programmer, given Brad's proven expertise, I believe he is well-suited for the project.”

“It seems to me that this solution is appropriate at this time in the project’s development..”

“If this can be done, it can probably speed up FreeCAD considerably. I don't have the technical knowledge to know if it's feasible and sensible, but Brad has worked on TNP and if someone can do this he sounds like a good candidate. I'm rather in favor of this.”

"I definitely support an effort to parallelize any part of FC possible."

**Comment by chennes** [comment]
Questions/comments that came up during the FPA vote:

> I am struck by the absence of "document all current uses of multiprocessing/multithreading in the existing code base".
> 
> I also didn't see anything along the lines of "determine where in the application slow downs occur and analyze if m/m would improve things", but I guess this could fit into 7)Preliminary profiling to generate baseline and estimate speedup.
> 
> These 2 seem to me to be prerequisites before investigating and picking solutions.
> 
> What "other projects" are envisioned for 12) Technical documentation for use of multiprocessing / multithreading in other projects?
> 
> Phases and weeks are used interchangeably and deliverables are not linked to phases. Are we paying for effort, or is there a deliverable for each phase/payment?

**Comment by bgbsww** [comment]
Absence of documenting current uses:  That is for the most part already done in my replies above, but will certainly be incorporated into the recommendations.

Determine where slow downs occur:  I am asserting with this request that a significant slowdown occurs in the recompute -> execute operation, and this is in fact one of, if not the single most significant slowdown to target initially.  Quantifying this is certainly within the profiling item.  I have spent some time profiling the application to determine where the slowdowns as part of the research required to even submit this, so I'm confident this is the one to target.

Other projects:  I don't know.  After this exercise completes, there will be a next most prevalent slowdown to investigate.  That isn't in scope here.  As a wild guess I would expect certain specific long running operations triggered by the GUI and needing user feedback to come up.  Whether there are systemic as opposed to individual speedups required is TBD.  A systemic one would be the Python GIL, but I don't see evidence that it is the chief culprit yet, which is why it isn't included here.

Deliverables:  There are 12 deliverables listed above.  Those are the 12 phases.  With the exception of #6 each of them has a visible output associated with it.  They are grouped for payment; for example, the 3 reports plus the approach to Document::getDependency list constitute the first one.  The timeline links the 12 deliverables to how long I think it will take for each one (11 weeks total, barring delays as noted by Yorik in his question above).  Apologies that this isn't clearer.

**Comment by prokoudine** [comment]
> I have spent some time profiling the application to determine where the slowdowns are as part of the research required to even submit this, so I'm confident this is the one to target.

Not sure if I'm allowed to comment here (not being part of the technical review committee), but it looks like presenting a short convincing overview of those findings would be really welcome. As in "this is what I profiled, this is how I did it, and this is what I discovered".

**Comment by bgbsww** [comment]
That's a fair point, but for me it starts to cross the line into even more speculative work that has not been funded.  If we want to proceed on this, I'm happy to do the work to summarize the evidence; if not than I want to cut my losses here.

We could also make this slower and more expensive by carving off a hunk of work to prove these points in advance; I'm doing my best here to preclude that as it doesn't seem efficient to me in this case.  I do appreciate the suggestion.

To try to explain by example:  I spent days investigating to refute the assertion that CLion could not be made to do simultaneous debugging and discovered the answer by digging deep into the implementation of all the python debuggers as pre-work to be able to make [grant proposal #13.](https://github.com/FreeCAD/FPA-grant-proposals/issues/13)  The actual [writeup work](https://github.com/FreeCAD/DevelopersHandbook/compare/main...bgbsww:DevelopersHandbook:bgbsww-DebuggerInstructions) and [PRs](https://github.com/FreeCAD/FreeCAD/pull/16850) and [side fixes](https://github.com/FreeCAD/FreeCAD/pull/16845) is actually only the downslope after figuring out the problem, and the overall project far exceeds the grant request.

This is similar, and matches the pattern of the TNP grant requests as well.  There is a mountain of invisible work here that I'm not requesting funding on.  I can make all of those steps more visible, but it will involve a long sequence of requests for smaller research grants with no tangible deliverables.  The phasing here is designed to allow the FPA to see tangible progress and evaluate as the work progresses to minimize the risk.

**Comment by prokoudine** [comment]
If the technical committee doesn't feel convinced that recomputation is indeed the bottleneck to focus on, one potential approach here would be doing a complete research as a first separate grant (which, I believe, would also include steps 1-3 at least) and then filing subsequent grant applications for further phases. That would minimize the amount of the prerequisite unpaid work for you while boosting committee's confidence.

**Comment by bgbsww** [comment]
Again, fair and a reasonable idea.  Unfortunately it greatly raises *my* risk, increases the amount of proposal review work necessary, and substantially increases the timeline.  I need some assurance that there is a commitment to keeping the work flowing at a rapid pace, or it multiplies the difficulty due to having to pick it up and put it back down, and makes other prospects more attractive.  Thus the preference for one large project.  If it isn't clear, I'm perfectly happy with placing abort points.  But splitting it piecemeal makes it much less attractive to consider.

**Comment by chennes** [comment]
The FPA supported funding this project, but due to [@bgbsww's death](https://blog.freecad.org/2024/10/08/rip-bradley-bgbsww-mclean/) it cannot be continued.

**Comment by mnesarco** [comment]
I left this comment here for future reference if someone is interested.

I have experience running a complex graph of OCCT operations in parallel. I don't know how **recompute** is implemented in FreeCAD, but AFAIK it must be based on a simple DAG. On any DAG we can get the **topological generations** and then evaluate each generation in sequence, but inside each generation we can evaluate all nodes in parallel. As every node in the same generation has fully solved dependencies and does not affect nodes in the same generation, there are no concurrent modifications.

I cannot take this project for FreeCAD now, but here are some useful references:

- https://en.wikipedia.org/wiki/Topological_sorting
- https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.dag.topological_generations.html
- https://www.rustworkx.org/dev/apiref/rustworkx.topological_generations.html


![image](https://github.com/user-attachments/assets/4116aa6a-f120-490a-a6b8-86a666a61ef3)
`image source: https://www.researchgate.net/figure/Example-view-of-topological-generations-layering-on-randomly-generated-DAGs_fig9_380601446`
